console-framework-client/src/main/java/io/axoniq/console/framework/client/AxoniqConsoleRSocketClient.kt

```diff
@@ -210,28 +188,99 @@ class AxoniqConsoleRSocketClient(
     }
+
+    private inner class HeartbeatOrchestrator : ClientSettingsObserver {
+        private var heartbeatSendTask: ScheduledFuture<*>? = null
+        private var heartbeatCheckTask: ScheduledFuture<*>? = null
+        private var lastReceivedHeartbeat = Instant.now()
+
+        init {
+            registrar.registerHandlerWithoutPayload(Routes.Management.HEARTBEAT) {
+                logger.debug("Received heartbeat from AxonIQ Console. Last one was: {}", lastReceivedHeartbeat)
+                lastReceivedHeartbeat = Instant.now()
+                lastReceivedHeartbeat
+            }
+        }
+
+        override fun onConnectedWithSettings(settings: ClientSettings) {
+            lastReceivedHeartbeat = Instant.now()
+            this.heartbeatSendTask = executor.scheduleWithFixedDelay(
+                    { sendHeartbeat().subscribe() },
+                    0,
+                    settings.heartbeatInterval,
+                    TimeUnit.MILLISECONDS
+            )
+
+            this.heartbeatCheckTask = executor.scheduleWithFixedDelay(
+                    { checkHeartbeats(settings.heartbeatTimeout) },
+                    0,
+                    1000,
+                    TimeUnit.MILLISECONDS
+            )
+
+        }
+
+        override fun onDisconnected() {
+            logger.info("Disconnected, stopping heartbeat tasks")
+            this.heartbeatSendTask?.cancel(true)
+            this.heartbeatCheckTask?.cancel(true)
+        }
+
+
+        private fun checkHeartbeats(heartbeatTimeout: Long) {
+            if (lastReceivedHeartbeat < Instant.now().minusMillis(heartbeatTimeout)) {
+                logger.info("Haven't received a heartbeat for {} seconds from AxonIQ Console. Reconnecting...", ChronoUnit.SECONDS.between(lastReceivedHeartbeat, Instant.now()))
+                disposeCurrentConnection()
+            }
+        }
+
+        private fun sendHeartbeat(): Mono<Payload> {
+            return rsocket
+                    ?.requestResponse(encodingStrategy.encode("", createRoutingMetadata(Routes.Management.HEARTBEAT)))
+                    ?.doOnSuccess {
+                        logger.debug("Heartbeat successfully sent to AxonIQ Console")
+                    }
+                    ?: Mono.empty()
+        }
+    }
+
+    private fun retrieveSettings(): Mono<ClientSettings> {
+        return rsocket!!
+                .requestResponse(encodingStrategy.encode("", createRoutingMetadata(Routes.Management.SETTINGS)))
+                .map {
+                    encodingStrategy.decode(it, ClientSettings::class.java)
+                }
+                .doOnError {
+                    if (it.message?.contains("Access Denied") == true) {
+                        logger.info("Was unable to send call to AxonIQ Console since authentication was incorrect!")
+                    }
+                }
+                .doOnSuccess {
+                    logger.debug("Heartbeat successfully sent to AxonIQ Console")
```

Comment:
Nit: This step is, strictly speaking, not the heartbeat, right?

Reply:
Correct, Removing the statement